### ISSUE TITLE
[java] MissingStaticMethodInNonInstantiatableClass: false positive with builder pattern

### ISSUE THREAD

**Issue by SaschaRiemer** [pmd/pmd#5068]
**Affects PMD Version:** 
7.2.0

**Rule:**
MissingStaticMethodInNonInstantiatableClass
https://docs.pmd-code.org/latest/pmd_rules_java_errorprone.html#missingstaticmethodinnoninstantiatableclass

**Description:**
Under certain circumstances, classes are identified as non-instantiatable, and then trigger the rule, even though the class can in fact be instantiated.

**Code Sample demonstrating the issue:**

```java
public class MyClass {

    private MyClass() {
    }

    public static class Builder {

        private Builder() {

        }

        public static Builder builder() {
            return new Builder();
        }

        public MyClass build() {
            return new MyClass();
        }

    }

}
```

**Expected outcome:**

PMD reports a violation at line 1, but that's wrong. That's a false positive. The class can be instantiated with `Builder.builder().build()`.

**Running PMD through:** *Gradle*

**Comment by jbisotti** [comment]
I am encountering the same false-positive when trying to upgrade from 6.55.0 to 7.5.0. We have several classes just like the one in the description above; this is preventing us from upgrading or forcing us to disable a rule we've used for ages. Any update on the likelihood this will be addressed in the near future?
